docker-compose-rule-junit4/src/main/java/com/palantir/docker/compose/EventEmitter.java

```diff
@@ -0,0 +1,134 @@
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class EventEmitter {
+    private static final Logger log = LoggerFactory.getLogger(EventEmitter.class);
+
+    private List<EventConsumer> eventConsumers;
+
+    EventEmitter(List<EventConsumer> eventConsumers) {
+        this.eventConsumers = eventConsumers;
+    }
+
+    interface CheckedRunnable {
+        void run() throws InterruptedException, IOException;
+    }
+
+    public void pull(CheckedRunnable runnable) throws IOException, InterruptedException {
+        emitThrowing(runnable, PullImagesEvent.FACTORY);
+    }
+
+    public void build(CheckedRunnable runnable) throws IOException, InterruptedException {
+        emitThrowing(runnable, BuildEvent.FACTORY);
+    }
+
+    public void up(CheckedRunnable runnable) throws IOException, InterruptedException {
+        emitThrowing(runnable, UpEvent.FACTORY);
+    }
+
+    public void waitingForServices(CheckedRunnable runnable) throws IOException, InterruptedException {
+        emitThrowing(runnable, WaitForServicesEvent.FACTORY);
+    }
+
+    public void shutdown(CheckedRunnable runnable) throws IOException, InterruptedException {
+        emitThrowing(runnable, ShutdownEvent.FACTORY);
+    }
+
+    public Consumer<Cluster> userClusterWait(ClusterWait clusterWait)  {
+        return clusterWait(ClusterWaitType.USER, clusterWait);
+    }
+
+    public Consumer<Cluster> nativeClusterWait(ClusterWait clusterWait) {
+        return clusterWait(ClusterWaitType.NATIVE, clusterWait);
+    }
+
+    private Consumer<Cluster> clusterWait(
+            ClusterWaitType clusterWaitType,
+            ClusterWait clusterWait) {
+        AtomicReference<Optional<Set<String>>> recordedServiceNames = new AtomicReference<>(Optional.empty());
+
+        Consumer<Cluster> recordingClusterWait = cluster -> {
+            RecordingCluster recordingCluster = new RecordingCluster(cluster);
+            try {
+                clusterWait.waitUntilReady(recordingCluster);
+            } finally {
+                recordedServiceNames.set(Optional.of(recordingCluster.recordedContainerNames()));
+            }
+        };
+
+
+        return cluster -> emitNotThrowing(
+                () -> recordingClusterWait.accept(cluster),
+                ClusterWaitEvent.factory(
+                        () -> recordedServiceNames.get().orElseThrow(
+                                () -> new IllegalStateException("Recorded service names have not yet been computed")),
+                        clusterWaitType));
+    }
+
+    private void emitNotThrowing(CheckedRunnable runnable, TaskEvent.Factory factory) {
+        try {
+            emitThrowing(runnable, factory);
+        } catch (InterruptedException | IOException e) {
+            Throwables.propagate(e);
+        }
+    }
+
+    private void emitThrowing(CheckedRunnable runnable, TaskEvent.Factory factory)
```

Comment:
wrap

Reply:
I think this is tricky as we call out `emit`, which uses the a field in this class. Unless we were to pass the `emit` function to `wrap`